app/src/main/java/net/gsantner/opoc/ui/SearchEngine.java

```diff
@@ -0,0 +1,440 @@
+package net.gsantner.opoc.ui;
+
+import android.app.Activity;
+import android.os.AsyncTask;
+import android.support.design.widget.Snackbar;
+import android.view.View;
+
+import net.gsantner.opoc.util.Callback;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.regex.Pattern;
+
+@SuppressWarnings("WeakerAccess")
+
+public class SearchEngine {
+    public static boolean isSearchExecuting = false;
+    public final static Pattern[] defaultIgnoredDirs = {
+            Pattern.compile("^\\.git$"),
+            Pattern.compile(".*[Tt]humb.*")
+    };
+    public final static Pattern[] defaultIgnoredFiles = {
+
+    };
+
+    public static class Config {
+        private final List<Pattern> _ignoredRegexDirs;
+        private final List<String> _ignoredExactDirs;
+        private final List<Pattern> _ignoredRegexFiles;
+        private final List<String> _ignoredExactFiles;
+
+        public final boolean isRegexQuery;
+        public final boolean isCaseSensitiveQuery;
+        public final String query;
+        public final File rootSearchDir;
+        public final boolean isSearchInContent;
+        public final boolean isShowResultOnCancel;
+        public final Integer maxSearchDepth;
+        public final List<String> ignoredDirectories;
+        public final List<String> ignoredFiles;
+
+        public Config(final File rootSearchDir, String query, final boolean isShowResultOnCancel,
+                      final Integer maxSearchDepth, final List<String> ignoredDirectories, final List<String> ignoredFiles,
+                      final boolean isRegexQuery, final boolean isCaseSensitiveQuery, final boolean isSearchInContent) {
+
+            this.rootSearchDir = rootSearchDir;
+            this.isSearchInContent = isSearchInContent;
+            this.isShowResultOnCancel = isShowResultOnCancel;
+            this.maxSearchDepth = maxSearchDepth;
+            this.ignoredDirectories = ignoredDirectories;
+            this.ignoredFiles = ignoredFiles;
+            this.isRegexQuery = isRegexQuery;
+            this.isCaseSensitiveQuery = isCaseSensitiveQuery;
+
+            this._ignoredExactDirs = new ArrayList<String>();
+            this._ignoredRegexDirs = new ArrayList<Pattern>();
+            splitRegexExactFiles(this.ignoredDirectories, this._ignoredExactDirs, this._ignoredRegexDirs);
+
+            this._ignoredExactFiles = new ArrayList<String>();
+            this._ignoredRegexFiles = new ArrayList<Pattern>();
```

Comment:
* Please use `_member = parameter`
* not `this._member = parameter`
* not `this.member = parameter`